src/Core/Variables/StandardVariableProvider.php

```diff
@@ -380,10 +398,44 @@ protected function isExtractableThroughAsserter($subject, $propertyName)
      */
     protected function extractThroughAsserter($subject, $propertyName)
     {
-        if (method_exists($subject, 'is' . ucfirst($propertyName))) {
-            return call_user_func_array([$subject, 'is' . ucfirst($propertyName)], []);
+        $className = $this->studyObject($subject);
+        $upperCasePropertyName = ucfirst($propertyName);
+        if (static::$gettersByClassName[$className]['is' . $upperCasePropertyName] ?? false) {
+            return call_user_func_array([$subject, 'is' . $upperCasePropertyName], []);
         }
 
-        return call_user_func_array([$subject, 'has' . ucfirst($propertyName)], []);
+        return call_user_func_array([$subject, 'has' . $upperCasePropertyName], []);
+    }
+
+    /**
+     * Studies an object and fills the static internal cache of getter method awareness.
+     * The method must be called when detecting an accessor for an object and has a very
+     * particular reason for existing:
+     *
+     * - method_exists() will return TRUE even for protected methods
+     * - is_callable() will not, but is far too greedy and returns TRUE for *any* method
+     *   name if __call is defined on the class.
+     * - disregarding Reflection which is horribly slow, get_class_methods is the only
+     *   method which returns a list of public methods.
+     * - but repeatedly calling get_class_methods() *AND* in_array() is not a sensible
+     *   solution.
+     * - therefore, an internal cached array is built by class name which allows a null-
+     *   coalesce expression to determine if a getter method exists *AND* is public *AND*
+     *   is not going to be handled by __call.
+     *
+     * For ease of use (to avoid repeating get_class calls) the method returns the class
+     * name of the instance as returned by get_class($instance).
+     *
+     * @param object $instance
+     * @return string
+     */
+    protected function studyObject($instance)
```

Comment:
I'd name this `populateGettersByClassName()` instead to increase the readability of the code. I understand that you don't directly return the map to reduce memory usage but right now the relation of this method and the static property is unclear without analyzing the code.

Reply:
Good choice.